### DISCUSSION TITLE
Increasing resolution when performing screenshots

### DISCUSSION THREAD

**Discussion by GonzaloSaez** [pyvista/pyvista#5938] [Q&A]
## Description

Hi,

I'm willing to create a "high-quality" png file of a 2D case. In paraview it is possible to set the **magnification** parameter in order to increase the quality of the png figures. To try perform this operation, I first select the 2D mode for the plotter

```python
p.enable_parallel_projection()
p.enable_image_style()
p.view_xy()
```

Add then I add a callback to perform my screenshot

```python
def screenshot():
    print("Window size ", p.window_size)
    p.screenshot("try.png", transparent_background=True, window_size=[2560,int(2560*p.window_size[1]/p.window_size[0])])
    print("Camera position ", p.camera_position)
p.add_key_event("s", screenshot)
```
When pressing the 's' key, I will obtain a screenshot with bigger window_size than the default value.
The problem with this approach is that the windows size varies, and I can't obtain a png of the zone of interest. The camera position seems to remain the same, although the picture is not the same if I increase the window_size

I have also tried using the save_graphic option

```python
p.save_graphic("try.pdf")
```

The pdf files seem to be heavy enough to represent vectorised data but the images are really blury in my opinion.

Is there any way in pyvista to increase the magnification and obtain higher quality png files?

## Example Data
- Original picture

![try](https://user-images.githubusercontent.com/11050889/72359882-62269200-36ef-11ea-8d93-8a9e8b646f0f.png)

- Changing window_size

![try](https://user-images.githubusercontent.com/11050889/72359803-3c00f200-36ef-11ea-9795-e2473ec1c1b1.png)

- Saving to pdf

[try.pdf](https://github.com/pyvista/pyvista-support/files/4060108/try.pdf)

Thank you very much for your help,

Gonzalo

**Comment by banesullivan** [comment]
I think the issues are arising from how the renderer is limited by the max resolution of your screen. When the plotters size changes and the height is greater than the screens resolution, the width will still expand and you'll end up with a different window shape and thus a different view of the ROI.

I'm not sure how we can address this. 

Perhaps do not automatically change the window size, but just maximize the window size.

I'll see if we can come up with a better routine for saving screenshots at a higher resolution.

**Comment by banesullivan** [comment]
We may need to set it up to do off-screen rendering during the screenshot phase

**Comment by GonzaloSaez** [comment]
Hi @banesullivan 

Thank you for the quick answer, I fixed this issue by using:

```python
p.ren_win.SetOffScreenRendering(1)
```

Before calling the screenshot function. The main pyvista window is lost after calling that function, but I can live with it.

**Comment by banesullivan** [comment]
Can you set it back on after the screenshot? And maybe reset the window size?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by GonzaloSaez** [comment]
Hi @banesullivan,

Following the vtk test :

https://gitlab.kitware.com/vtk/vtk/blob/107d84fe94c7e3122dac27621df4ef9f8a1e299e/Rendering/Core/Testing/Cxx/TestToggleOSWithInteractor.cxx

I have added the following lines of code to my script:

```python
hsize=2560
p.ren_win.OffScreenRenderingOn()
p.enable_anti_aliasing()
p.screenshot("try.png", transparent_background=True, window_size=[hsize, int(hsize*p.window_size[1]/p.window_size[0])])
p.ren_win.SetSize([1000, 800])
p.ren_win.OffScreenRenderingOff()
p.ren_win.Render()
```

But the screen doesn't reappear.

**Comment by e-dub** [comment]
I am trying to export a high quality PNG and came across this issue and also [this](https://github.com/pyvista/pyvista/issues/105).  I do not get either procedure. Is there an example somewhere for high quality exports?

I am currently using the following command
`
plotter.show(screenshot="Mesh.png")`

which is providing a very pixelly result.
